File: src/main/java/cz/grunwjir/csasdemo/branch/service/BranchServiceImpl.java
package cz.grunwjir.csasdemo.branch.service;

import cz.grunwjir.csasdemo.Page;
import cz.grunwjir.csasdemo.branch.model.Branch;
import cz.grunwjir.csasdemo.csas.client.BranchData;
import cz.grunwjir.csasdemo.csas.client.CsasClient;
import cz.grunwjir.csasdemo.csas.client.CsasResponse;
import cz.grunwjir.csasdemo.gmaps.service.GMapsService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;


/**
 * Implementation of {@link BranchService}.
 *
 * @author Jiri Grunwald
 */
@Slf4j
@RequiredArgsConstructor
@Service
public class BranchServiceImpl implements BranchService {

    private final GMapsService gMapsService;
    private final CsasClient csasClient;

    public Page<Branch> searchBranches(String query, Integer size, Integer page) {
        log.info("Searching branch with review: {} (size: {}, page: {})", query, size, page);
        CsasResponse<BranchData> response = csasClient.searchBranches(query, size, page);

        List<Branch> branches = response.getItems().stream()
                .map(data -> {
                    Branch branch = mapBranch(data);

                    // add branch rating
                    branch.setGoogleMapsRating(gMapsService.getPlaceRating(branch.getName(),
                            branch.getAddress().getStreet(), branch.getAddress().getCity()));

                    return branch;
                })
                .collect(Collectors.toList());

        log.info("Total: {} branches were found with query text: {}", response.getTotalItemCount(), query);

        return Page.<Branch>builder()
                .items(branches)
                .pageNumber(response.getPageNumber())
                .pageCount(response.getPageCount())
                .pageSize(response.getPageSize())
                .totalItemCount(response.getTotalItemCount())
                .build();
    }

    private Branch mapBranch(BranchData data) {
        Branch branch = new Branch();
        branch.setId(data.getId());
        branch.setName(data.getName());
        branch.setManager(data.getManagerName());
        branch.setPhones(data.getPhones());
        branch.setEquipments((data.getEquipments() != null) ? data.getEquipments().stream()
                .map(BranchData.Equipment::getName)
                .collect(Collectors.toList()) : List.of());
        branch.setEmail(data.getEmail());
        branch.setAddress(new Branch.Address(data.getAddress(), data.getCity(), data.getPostCode()));

        return branch;
    }
}

File: src/main/kotlin/cz/grunwjir/csasdemo/branch/service/BranchServiceImpl.kt
package cz.grunwjir.csasdemo.branch.service

import cz.grunwjir.csasdemo.Page
import cz.grunwjir.csasdemo.branch.model.Branch
import cz.grunwjir.csasdemo.csas.client.BranchData
import cz.grunwjir.csasdemo.csas.client.CsasClient
import cz.grunwjir.csasdemo.csas.client.CsasResponse
import cz.grunwjir.csasdemo.gmaps.service.GMapsService
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

/**
 * Implementation of [BranchService].
 *
 * @author Jiri Grunwald
 */
@Service
class BranchServiceImpl(private val gMapsService: GMapsService, private val csasClient: CsasClient) : BranchService {

    companion object {
        private val log: Logger = LoggerFactory.getLogger(BranchServiceImpl::class.java)
    }

    override fun searchBranches(query: String, size: Int?, page: Int?): Page<Branch> {
        log.info("Searching branch with review: {} (size: {}, page: {})", query, size, page)
        val csasResponse: CsasResponse<BranchData> = csasClient.searchBranches(query, size, page)

        return csasResponse.items
            .map { mapBranch(it, gMapsService.getPlaceRating(it.name, it.address, it.city)) }
            .let {
                Page(
                    items = it,
                    pageNumber = csasResponse.pageNumber,
                    pageCount = csasResponse.pageCount,
                    pageSize = csasResponse.pageSize,
                    totalItemCount = csasResponse.totalItemCount
                )
            }
            .also { log.info("Total: {} branches were found with query text: {}", csasResponse.totalItemCount, query) }
    }

    private fun mapBranch(data: BranchData, googleMapsRating: Float?): Branch {
        return Branch(
            id = data.id,
            name = data.name,
            manager = data.managerName ?: "",
            phones = data.phones ?: emptyList(),
            email = data.email ?: "",
            equipments = data.equipments?.map { it.name } ?: emptyList(),
            address = Branch.Address(
                street = data.address,
                city = data.city,
                postalCode = data.postCode
            ),
            googleMapsRating = googleMapsRating
        )
    }
}